Add `pyjion`

dhirschfeld opened the pull request:
<!--
Thank you very much for putting in this recipe PR!

This repository is very active, so if you need help with
a PR or once it's ready for review, please let the right people know.
There are language-specific teams for reviewing recipes.

Currently available teams are:
- python `@conda-forge/help-python`
- python/c hybrid `@conda-forge/help-python-c`
- r `@conda-forge/help-r`
- java `@conda-forge/help-java`
- nodejs `@conda-forge/help-nodejs`
- c/c++ `@conda-forge/help-c-cpp`
- perl `@conda-forge/help-perl`
- Julia `@conda-forge/help-julia`
- ruby `@conda-forge/help-ruby`

If your PR doesn't fall into those categories please contact
the full review team `@conda-forge/staged-recipes`.

Due to GitHub limitations first time contributors to conda-forge are unable
to ping these teams. You can [ping the team][1] using a special command in
a comment on the PR to get the attention of the `staged-recipes` team. You can
also consider asking on our [Gitter channel][2] if your recipe isn't reviewed promptly.

[1]: https://conda-forge.org/docs/maintainer/infrastructure.html#conda-forge-admin-please-ping-team
[2]: https://gitter.im/conda-forge/conda-forge.github.io
-->

Checklist
- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/5eddbd7fc9d1502169089da06c3688d9759be978/recipes/example/meta.yaml#L64-L73) for an example).
- [x] Source is from official source.
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- [x] If static libraries are linked in, the license of the static library is packaged.
- [x] Build number is 0.
- [ ] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [x] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I wanted to let you know that I linted all conda-recipes in your PR (```recipes/pyjion```) and found some lint.

Here's what I've got...


For **recipes/pyjion**:

 * Non noarch packages should have python requirement without any version constraints.
 * Non noarch packages should have python requirement without any version constraints.

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipes/pyjion```) and found it was in an excellent condition.

dhirschfeld commented:
### Note!

This is using a `git_url` as the [sources aren't published on `pypi`](https://pypi.org/project/pyjion/#files) and the build requires submodules to be present.

dhirschfeld commented:
IIUC the `pyjion` repo builds using the `dotnet` submodule:
* https://github.com/tonybaloney/Pyjion/blob/develop/main/.gitmodules

I'm checking with the author what that means in terms of also distributing the [.NET license](https://github.com/dotnet/runtime/blob/main/LICENSE.TXT)

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I wanted to let you know that I linted all conda-recipes in your PR (```recipes/pyjion```) and found some lint.

Here's what I've got...


For **recipes/pyjion**:

 * The test section contained an unexpected subsection name. commnds is not a valid subsection name.

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipes/pyjion```) and found it was in an excellent condition.

dhirschfeld commented:
It *looks* like CI is "passing" because all the builds are being skipped?! 🤔

dhirschfeld commented:
I've removed the *draft* status in case the solution here is to simply merge and debug in the feedstock.

Will see what the experts in @conda-forge/staged-recipes think...

chrisburr commented on recipes/pyjion/meta.yaml:
```diff
@@ -0,0 +1,68 @@
+
+requirements:
+  build:
+    - cmake
```
Do you not need a compiler?

chrisburr replied:
It has `dotnet` 🤦

dhirschfeld replied:
It does, but I'm not sure if it also requires a C or C++ compiler. From the languages, I think it probably does:

![image](https://user-images.githubusercontent.com/881019/152682226-b8b6ddd0-12f5-4668-921c-6347009be25e.png)

https://github.com/tonybaloney/Pyjion

dhirschfeld replied:
![image](https://user-images.githubusercontent.com/881019/152682495-d05690b3-b738-4707-8e20-71b90f40fc9e.png)
https://pyjion.readthedocs.io/en/latest/development.html#compiling-from-source

Suggests you need VS2019 (on Windows I guess). Fortunately, it seems VS2019 support was added in [`scikit-build` 0.12.0](https://github.com/scikit-build/scikit-build/releases/tag/0.12.0) whcih is [available on CF](https://github.com/conda-forge/scikit-build-feedstock).

dhirschfeld replied:
I'll unresolve so @tonybaloney doesn't miss this question. Otherwise, I imagine CI will tell us pretty quickly if a compiler is missing!

chrisburr commented on recipes/pyjion/meta.yaml:
```diff
@@ -0,0 +1,68 @@
+    - git
+    - ninja
+  host:
+    - python
```
For testing you can add this to `host` an `run` (remove it before merging):
```suggestion
    - python 3.10.* *_cpython
```

dhirschfeld commented:
It seems `cmake` isn't using `ninja`. Looks like I need to set `CMAKE_MAKE_PROGRAM`. That'll have to be a problem for the morning though!
```
  Running command Building wheel for pyjion (pyproject.toml)


  --------------------------------------------------------------------------------
  -- Trying "Unix Makefiles" generator
  --------------------------------
  ---------------------------
  ----------------------
  -----------------
  ------------
  -------
  --
  Not searching for unused variables given on the command line.
  CMake Error: CMake was unable to find a build program corresponding to "Unix Makefiles".  CMAKE_MAKE_PROGRAM is not set.  You probably need to select a different build tool.
  -- Configuring incomplete, errors occurred!
  See also "/home/conda/staged-recipes/build_artifacts/pyjion_1644155830840/work/_cmake_test_compile/build/CMakeFiles/CMakeOutput.log".
  --
  -------
  ------------
  -----------------
  ----------------------
  ---------------------------
  --------------------------------
  -- Trying "Unix Makefiles" generator - failure
  --------------------------------------------------------------------------------

  ********************************************************************************
  scikit-build could not get a working generator for your system. Aborting build.
  
  × Building wheel for pyjion (pyproject.toml) did not run successfully.
  │ exit code: 1
  ╰─> See above for output.
```

dhirschfeld commented:
The env var seemed to get passed through but still wasn't used:
```
UserWarning: The environment variable 'CMAKE_GENERATOR' is being passed through with value 'Ninja'.
If you are splitting build and test phases with --no-test, please ensure that this value is also set similarly at test time.
```

dhirschfeld commented:
Odd MacOS error:
```
==> /Users/runner/mambaforge/bin/git describe --tags --dirty <==

1.2.6

fatal: iconv_open(UTF-8,UTF-8-MAC) failed, but needed:
    precomposed unicode is not supported.
    If you want to use decomposed unicode, run
    "git config core.precomposeunicode false"

fatal: 'git status --porcelain=2' failed in submodule CoreCLR
```
***xref:*** https://github.com/conda-forge/git-feedstock/issues/50#issuecomment-552119855

dhirschfeld commented on recipes/pyjion/meta.yaml:
```diff
@@ -0,0 +1,73 @@
+    echo '####################################################################'
+    echo "${CMAKE_MAKE_PROGRAM}"
+    echo '####################################################################'
+    {{ PYTHON }} -m pip install . -vv
```
It looks like this is working:

```
####################################################################
$BUILD_PREFIX/bin/ninja
####################################################################
```
...but it still doesn't appear to be being tried:
```
  -- Trying "Unix Makefiles" generator - failure
  --------------------------------------------------------------------------------

  ********************************************************************************
  scikit-build could not get a working generator for your system. Aborting build.
```

dhirschfeld commented:
That seemed to get a lot further:
```
  -- Trying "Unix Makefiles" generator - success
  --------------------------------------------------------------------------------

  Configuring Project
    Working directory:
      /home/conda/staged-recipes/build_artifacts/pyjion_1644194641329/work/_skbuild/linux-x86_64-3.10/cmake-build
    Command:
      cmake /home/conda/staged-recipes/build_artifacts/pyjion_1644194641329/work -G 'Unix Makefiles' -DCMAKE_INSTALL_PREFIX:PATH=/home/conda/staged-recipes/build_artifacts/pyjion_1644194641329/work/_skbuild/linux-x86_64-3.10/cmake-install -DPYTHON_EXECUTABLE:FILEPATH=/home/conda/staged-recipes/build_artifacts/pyjion_1644194641329/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_plac/bin/python -DPYTHON_VERSION_STRING:STRING=3.10.2 -DPYTHON_INCLUDE_DIR:PATH=/home/conda/staged-recipes/build_artifacts/pyjion_1644194641329/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_plac/include/python3.10 -DPYTHON_LIBRARY:FILEPATH=/home/conda/staged-recipes/build_artifacts/pyjion_1644194641329/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_plac/lib/libpython3.10.so -DSKBUILD:INTERNAL=TRUE -DCMAKE_MODULE_PATH:PATH=/home/conda/staged-recipes/build_artifacts/pyjion_1644194641329/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_plac/lib/python3.10/site-packages/skbuild/resources/cmake -DCMAKE_BUILD_TYPE:STRING=Release

  -- The C compiler identification is GNU 9.4.0
  -- The CXX compiler identification is GNU 9.4.0
  -- Detecting C compiler ABI info
  -- Detecting C compiler ABI info - done
  -- Check for working C compiler: /home/conda/staged-recipes/build_artifacts/pyjion_1644194641329/_build_env/bin/x86_64-conda-linux-gnu-cc - skipped
  -- Detecting C compile features
  -- Detecting C compile features - done
  -- Detecting CXX compiler ABI info
  -- Detecting CXX compiler ABI info - done
  -- Check for working CXX compiler: /home/conda/staged-recipes/build_artifacts/pyjion_1644194641329/_build_env/bin/x86_64-conda-linux-gnu-c++ - skipped
  -- Detecting CXX compile features
  -- Detecting CXX compile features - done
  -- Enabling PIC
  -- Found PythonInterp: /home/conda/staged-recipes/build_artifacts/pyjion_1644194641329/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_plac/bin/python (found version "3.10.2")
  -- Found PythonLibs: /home/conda/staged-recipes/build_artifacts/pyjion_1644194641329/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_plac/lib/libpython3.10.so (found version "3.10.2")
  -- Using Python libraries for skbuild /home/conda/staged-recipes/build_artifacts/pyjion_1644194641329/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_plac/lib/libpython3.10.so
  -- Using Python headers for skbuild /home/conda/staged-recipes/build_artifacts/pyjion_1644194641329/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_plac/include/python3.10
  -- Using Python libraries /home/conda/staged-recipes/build_artifacts/pyjion_1644194641329/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_plac/lib/libpython3.10.so
  -- Using Python headers /home/conda/staged-recipes/build_artifacts/pyjion_1644194641329/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_plac/include/python3.10
  CMAKE_HOST_SYSTEM_PROCESSOR: x86_64
  CMAKE_SYSTEM_PROCESSOR: x86_64
  -- Enabling UNIX Patches
  -- Detected CPU arch x86_64
  -- Enabling AMD64
  -- Enabling AMD64 ABI
  -- Found .NET 6 in /home/conda/staged-recipes/build_artifacts/pyjion_1644194641329/_build_env/lib/dotnet/shared/Microsoft.NETCore.App/6.0.1
  -- Performing Test Weak Link MODULE -> SHARED (gnu_ld_ignore) - Success
  _modinit_prefix:PyInit_
  -- Configuring done
  -- Generating done
  -- Build files have been written to: /home/conda/staged-recipes/build_artifacts/pyjion_1644194641329/work/_skbuild/linux-x86_64-3.10/cmake-build
  [  5%] Building CXX object CMakeFiles/pyjionlib.dir/src/pyjion/absint.cpp.o
  [ 11%] Building CXX object CMakeFiles/pyjionlib.dir/src/pyjion/absvalue.cpp.o
  [ 17%] Building CXX object CMakeFiles/pyjionlib.dir/src/pyjion/intrins.cpp.o
  [ 23%] Building CXX object CMakeFiles/pyjionlib.dir/src/pyjion/jitinit.cpp.o
  [ 29%] Building CXX object CMakeFiles/pyjionlib.dir/src/pyjion/pycomp.cpp.o
  [ 35%] Building CXX object CMakeFiles/pyjionlib.dir/src/pyjion/pyjit.cpp.o
  [ 41%] Building CXX object CMakeFiles/pyjionlib.dir/src/pyjion/exceptionhandling.cpp.o
  [ 47%] Building CXX object CMakeFiles/pyjionlib.dir/src/pyjion/stack.cpp.o
  [ 52%] Building CXX object CMakeFiles/pyjionlib.dir/src/pyjion/codemodel.cpp.o
  [ 58%] Building CXX object CMakeFiles/pyjionlib.dir/src/pyjion/binarycomp.cpp.o
  [ 64%] Building CXX object CMakeFiles/pyjionlib.dir/src/pyjion/instructions.cpp.o
  [ 70%] Building CXX object CMakeFiles/pyjionlib.dir/src/pyjion/unboxing.cpp.o
  [ 76%] Building CXX object CMakeFiles/pyjionlib.dir/src/pyjion/pgc.cpp.o
  [ 82%] Building CXX object CMakeFiles/pyjionlib.dir/src/pyjion/base.cpp.o
  [ 88%] Building CXX object CMakeFiles/pyjionlib.dir/src/pyjion/objects/unboxedrangeobject.cpp.o
  [ 94%] Building CXX object CMakeFiles/pyjionlib.dir/src/pyjion/attrtable.cpp.o
  [ 94%] Built target pyjionlib
  [100%] Linking C shared module _pyjion.cpython-310-x86_64-linux-gnu.so
  [100%] Built target _pyjion

```
`...`
```
  Building wheel for pyjion (pyproject.toml): finished with status 'done'
  Created wheel for pyjion: filename=pyjion-1.2.6-cp310-cp310-linux_x86_64.whl size=494065 sha256=7e0a8ca93a02c2f3ad21ef8e8cf3a20e18770ee56ed23180318fd64a7a37cf5a
  Stored in directory: /tmp/pip-ephem-wheel-cache-1oowxlgu/wheels/35/80/25/fec4501f039d50efca7267742da79f83d902cbbc2b7ff4e098
Successfully built pyjion
Installing collected packages: pyjion

  changing mode of $PREFIX/bin/pyjion to 755
Successfully installed pyjion-1.2.6
Removed build tracker: '/tmp/pip-req-tracker-djgu53t2'
```
...but ultimately failed with:
```
ERROR (pyjion,lib/python3.10/site-packages/pyjion/_pyjion.cpython-310-x86_64-linux-gnu.so):
    $RPATH/libclrjit.so not found in packages, sysroot(s) nor the missing_dso_whitelist
    ...is this binary repackaging?
```

dhirschfeld commented:
Well, that seemed to work :tada:

dhirschfeld commented on recipes/pyjion/meta.yaml:
```diff
@@ -0,0 +1,81 @@
+    - '*/libclrjit.so'  # [linux]
+  script_env:
+    - CMAKE_GENERATOR=Ninja
+  script: |
```
It seems the `CMAKE_GENERATOR` variable is being passed through but not used?
```
####################################################################
CMAKE_GENERATOR=Unix Makefiles
CMAKE_MAKE_PROGRAM=$BUILD_PREFIX/bin/ninja
####################################################################
UserWarning: The environment variable 'CMAKE_GENERATOR' is being passed through with value 'Ninja'. 
```

dhirschfeld commented:
I've disabled `win` & `osx` for now as I'm only (initially) interested in `linux` support and it's probably easier to enable & debug the other platforms in the feedstock.

dhirschfeld commented:
The import "test" worked by pytest doesn't seem to recognise the `--opt-level` arg:
```
import: 'pyjion'
import: 'pyjion'
+ python -X dev -m pytest Tests/ -v --opt-level 1
ERROR: usage: __main__.py [options] [file_or_dir] [file_or_dir] [...]
__main__.py: error: unrecognized arguments: --opt-level 1
  inifile: None
  rootdir: /home/conda/staged-recipes/build_artifacts/pyjion_1644200273570/test_tmp

Tests failed for pyjion-1.2.6-py310haf19d99_0.tar.bz2 - moving package to /home/conda/staged-recipes/build_artifacts/broken
```

dhirschfeld commented:
I've enabled most tests here (IIUC the external ones require `distnorm3` which isn't yet available on CF). Normally CF rely on upstream testing and just do basic import tests of the CF package. In this case, with `pyjion` being a complex, compiled package I think running the test suite is useful to ensure the compilation of the CF package is correct and hasn't broken anything.

dhirschfeld commented:
The test-suite appears to pass on linux so I think this is ready for review/merge!

dhirschfeld commented:
PTAL @conda-forge/staged-recipes - I think this is good to merge!

dhirschfeld commented:
There's a lot of top-class, whack-a-mole debugging in this PR - I'd suggest squashing & merging! 😅

isuruf commented on recipes/pyjion/meta.yaml:
```diff
@@ -0,0 +1,79 @@
+  build:
+    - {{ compiler('cxx') }}
+    - cmake
+    - dotnet
```
Move this to host and remove `missing_dso_whitelist`

chrisburr commented on recipes/pyjion/meta.yaml:
```diff
@@ -0,0 +1,79 @@
+  version: {{ version }}
+
+source:
+  git_url: https://github.com/tonybaloney/Pyjion.git
+  git_tag: {{ version }}
```
`git_url` is strongly discouraged (the bot doesn't understand it, no hashing to make sure the tag doesn't change, ...)

```suggestion
  url: https://github.com/tonybaloney/Pyjion/archive/refs/tags/{{ version }}.tar.gz
  sha256: 1d76194037114d674eada64bea2c04ab276880cf5da4325be186e07d7ce700d2
```

dhirschfeld replied:
I am aware of that, however, this was done specifically because the build uses git submodules which IIUC are not included in the GitHub release archives.

dhirschfeld replied:
IIUC the only workaround is to include the submodule as a separate source download and to unpack it *into* the `pyjion` source folder.

You will need to find out the exact commit hash used and AFAIK this also isn't supported by the bot so it seems to add a lot of complexity and manual work for no gain?

dhirschfeld replied:
> *no hashing to make sure the tag doesn't change*

If the concern is mutable tag references you can reference the actual commit instead I think?

dhirschfeld replied:
There's a fair bit of discussion about what to do in the case of submodules - e.g.
* https://github.com/conda-forge/xgboost-feedstock/issues/22#issuecomment-475972221
* https://github.com/conda-forge/conda-forge.github.io/issues/619#issuecomment-507393863
* https://github.com/conda-forge/coolprop-feedstock/pull/8/files#r331113300

From my point of view, if the bot can't correctly update the submodule source, there isn't much point in bending over backwards to use it. Even if the submodule hash referred to a tag, you don't actually want the bot greedily updating the submodule if it hasn't been updated in the actual repo.

If the bot won't work either way, I'm happy to manually put in PRs to update this package and I think using the `git_url` will incur the lowest maintenance burden.

chrisburr replied:
Ah right, I hadn't realised submodules were being used. Yeah in that case I'm happy enough with it using the tag. I'd rather avoid risking the version and commit rev get out of sync.

dhirschfeld commented on recipes/pyjion/meta.yaml:
```diff
@@ -0,0 +1,79 @@
+{% set name = "pyjion" %}
+{% set version = "1.2.6" %}
```
![image](https://user-images.githubusercontent.com/881019/152985280-da753f0d-0efc-4ca4-bdae-93dbca79278f.png)

The commit hash (https://github.com/tonybaloney/Pyjion/commit/66a772d48825e2511b9c6e8bbcc739d84f84e800) could be used instead if there are concerns the tag could change?

dhirschfeld commented:
Also, thanks for the review @chrisburr and @isuruf! I'm happy to go with the multiple sources if that's the decision - I just wanted to explain my reasoning here.

isuruf commented:
> The dotnet SDK is a binary repackage so perhaps that's why it's required here? Investigating building dotnet from source is on my TODO list.

No, that's because `_pyjion.cpython-310-x86_64-linux-gnu.so` is built without the rpath.

dhirschfeld commented:
> *No, that's because _pyjion.cpython-310-x86_64-linux-gnu.so is built without the rpath.*

Ok. Is there a way to fix that, or should I just keep the `missing_dso_whitelist`?

dhirschfeld commented:
Ok, re-added `missing_dso_whitelist`. Build/tests passing. I think I've addressed all the review issues, let me know if you'd like any other changes...

dhirschfeld commented:
Ok, re-added `missing_dso_whitelist`. Build/tests passing. I think I've addressed all the review issues, let me know if you'd like any other changes...

dhirschfeld commented:
ping @chrisburr, @isuruf since you're most familiar with this PR. It would be great to get this merged if you're happy with it?

chrisburr commented on recipes/pyjion/meta.yaml:
```diff
@@ -0,0 +1,80 @@
+source:
+  git_url: https://github.com/tonybaloney/Pyjion.git
+  # https://github.com/tonybaloney/Pyjion/tree/1.2.6
+  git_rev: 66a772d48825e2511b9c6e8bbcc739d84f84e800
```
https://github.com/conda-forge/staged-recipes/pull/17937#discussion_r801656393
```suggestion
  git_rev: {{ version }}
```

dhirschfeld commented:
CI passed before so I'm happy for this to be merged.
```
No hosted parallelism has been purchased or granted
```

dhirschfeld commented:
ping!
